ADR/0033-apps-continuous-chaos-testing.md

```diff
@@ -0,0 +1,41 @@
+# 18. Continuous Chaos Testing of Apps in AppStudio
+
+Date: 2024-03-05
+
+## Status
+
+In consideration
+
+## Context
+
+The chaos engineering strategy enables users to discover potential causes of service degradation. It helps users understand their app behavior under unpredictable conditions, identify areas to harden, and utilize performance data points to size and tune their application to handle failures, thereby minimizing downtime.
+
+There are two approaches to chaos testing in the CI/CD pipeline.
+
+### Resilience based Chaos scenario
+
+These Chaos scenarios are expected to cause application failure. Example scenarios include simulating memory pressure, storage errors, killing random or dependent resources. The objective of these chaos test cases in the CI/CD pipeline is to assess whether the application is capable of mitigating and maintaining reliability.
+
+![Architecture diagram of Resilience based Chaos test scenario](../diagrams/ADR-0033/chaos-resilience.png "Architecture diagram of Resilience based Chaos test scenario")
+
+### SLA based Chaos scenario
+
+Test the resiliency of a application under turbulent conditions by running tests that are designed to disrupt while monitoring the application adaptability and performance:
+Establish and define your steady state and metrics - understand the behavior and performance under stable conditions and define the metrics that will be used to evaluate the application’s behavior. Then decide on acceptable outcomes before injecting chaos.
+Analyze the statuses and metrics of all components during the chaos test runs.
+Improve the areas that are not resilient and performant by comparing the key metrics and Service Level Objectives (SLOs) to the stable conditions before the chaos. For example: evaluating the API server latency or application uptime to see if the key performance indicators and service level indicators are still within acceptable limits.
+
+![Architecture diagram of SLA based Chaos test scenario](../diagrams/ADR-0033/chaos-sla.png "Architecture diagram of SLA based Chaos test scenario")
+
+
+### Glossary
+
+- krkn: Chaos testing framework: <https://github.com/krkn-chaos/krkn>
+
+## Decision
+
+The Knoflux user can run chaos tests such as Krkn as part of the IntegrationTestScenarios in the ephemeral environment and can gather Prometheus metrics from Thanos-querier-openshift-monitoring or Prometheus-k8s-openshift-monitoring endpoints.
+
+## Consequences
+
+The service account associated with the ephemeral environment will possess admin-level privileges to execute CRUD operations within the ephemeral namespace space. Using the service account token, it will be possible to authenticate and query Prometheus metrics from Thanos-querier-openshift-monitoring or Prometheus-k8s-openshift-monitoring endpoints.
```

Comment:
Which resources will the chaos test need CRUD access to? We should call them out here.

Also, I believe granting the SA access to query metrics will allow (with a little "know how") the end user to access all metrics collected by that prometheus instance, not just those produced for the temp namespace. Are we sure we want to grant such access? We haven't exposed prometheus metrics to end users yet and this isn't taking the least privileged approach.

Reply:
Hello @amisstea,
At present the serviceaccount associated with the ephemeral environment, seem to have all the required CURD access for pods,services, networkpolicies, deployments, replicastes, route.
So no additional modification would be necessary.
| **Resources**                      | **Non-Resource URLs** | **Resource Names** | **Verbs**                                                    |
|------------------------------------|-----------------------|--------------------|--------------------------------------------------------------|
| pods                               |         []            |       []           | [get list watch create delete deletecollection patch update] |
| services                           |         []            |       []           | [get create list watch delete deletecollection patch update] |
| networkpolicies.networking.k8s.io  |         []            |       []           | [create delete deletecollection patch update get list watch] |
| replicasets.apps                   |         []            |       []           | [create delete deletecollection patch update get list watch] |
| deployments.apps                   |         []            |       []           | [get list watch create delete deletecollection patch update] |
| routes                             |         []            |       []           | [create delete deletecollection get list patch update watch] |


If there exists a less privileged method for querying Prometheus metrics, I would be more than happy to provide support for such an implementation.Since it would not just benefit this ADR but also the  [0018-apps-continuous-perf-testing ](ADR/0018-apps-continuous-perf-testing.md)ADR related to performance.
The downside of not enabling the SA to query the Prometheus endpoints would be that we would miss out on executing the SLA-based Chaos scenario, but we should still be able to run the Resilience-based Chaos scenario on Konflux.

Reply:
I'm not aware of a way to make this less privileged. If I'm not mistaken, limiting prometheus access to the service account used in the krkn pod won't prevent an informed user from utilizing that SA in other workloads. @yftacherzog @avi-biton do you have any ideas?

@yogananth-subramanian please add a little more detail here about how this change will grant the service account unfettered access to query any metric within the cluster, not just those in the ephemeral namespace.

Reply:
What are the metrics that should be collected here? Is it some custom metrics generated by exporter on the same namespace or generic metrics generated by KSM and such?
Maybe the overhead of deploying an ephemeral prometheus instance inside the namespace could make sense?

Reply:
We discussed this problem during the arch call yesterday and reached a consensus that chaos tests should utilize ephemeral clusters (see #172) rather than ephemeral namespaces for better isolation and a more production-like testing environment. The end user will have cluster admin access on such clusters to configure RBAC permissions, prometheus instances, etc.

@yogananth-subramanian, at your convenience, please update this ADR to reflect the decision.

Reply:
@yogananth-subramanian have you had a chance to look into update here?

Reply:
The end user will have cluster-admin permissions on the ephemeral clusters so there's no need to call out the specific resource requirements. Resolving this thread.